### ISSUE TITLE
Retry Listener

### ISSUE THREAD

**Issue by kevinconaway** [rholder/guava-retrying#36]
What are your thoughts on adding a listener framework to the the Retryer?

My use case is that I would like to log each failed retry attempt but I don't see any way to inspect the results

I was thinking we could add a simple interface

```java
interface RetryListener {
    void attempt(Attempt attempt, int attemptNumber);
}
```
That should cover most use cases.  [Spring retry](https://github.com/spring-projects/spring-retry) has a different interface but I think just logging each `Attempt` should cover onClose/onError.

Happy to create a pull request for this if you're interested.

**Comment by kevinconaway** [comment]
Looks like something similar is proposed in https://github.com/rholder/guava-retrying/pull/21

**Comment by kevinconaway** [comment]
@rholder any thoughts here? It looks like this has come up a few times.

**Comment by rholder** [comment]
It has come up a few times. Would the PR in #21 handle your case, too?

**Comment by kevinconaway** [comment]
I think the implementation in this fork https://github.com/txtr/guava-retrying/commit/a13cc7d1e2e766cdcf74d1a5edbc117bd9989fe4 was closer to what I had in mind.

The `FailedAttemptEvent` concept doesn't make sense if you want to capture successful invocations too.

**Comment by rholder** [comment]
I would be in favor of a solution similar to what's in that fork. Feel free to open a PR and if you don't hear back from me on it just keep commenting until I respond.

**Comment by kevinconaway** [comment]
Hey @rholder any update on when this will be released?

Thanks!

**Comment by rholder** [comment]
It's getting closer to a release. I've been swamped lately, but hope to block out some time this weekend.

**Comment by kevinconaway** [comment]
Hi @rholder any news on 1.8 or 2.0 release?

Thank you

**Comment by kevinconaway** [comment]
Hi @rholder, is there an ETA when this will be released?

**Comment by rholder** [comment]
@kevinconaway This is for sure the next project I'm blocking out time to ship. I'm setting a personal deadline for getting a 2.0 out at the end of June (regardless of how many things don't make it in). Thanks for continuing to send me reminder feedback that there are things to do here because it helps me prioritize projects.

**Comment by rholder** [comment]
Your implementation has landed in the 2.0.0 release that just went out. I'd planned to eventually be able to register listeners specifically for failure events, too, but I would rather release this now since your functionality is complete and appears to be working delightfully. I've added a `@Beta` annotation to give a heads up to early adopters that the API is still subject to change moving forward in case we get that in there at a later date.

**Comment by rholder** [comment]
Thanks again for your contributions and relentless patience!
